ENH: str.extractall for several matches

tdhock opened the pull request:
For a series `S`, the excellent `S.str.extract` method returns the first match in each subject of the series:

``` python
>>> import re
>>> import pandas as pd
>>> import numpy as np
>>> data = {
...     'Dave': 'dave@google.com',
...     'multiple': 'rob@gmail.com some text steve@gmail.com',
...     'none': np.nan,
...     }
>>> pattern = r'''
... (?P<user>[a-z]+)
... @
... (?P<domain>[a-z]+)
... \.
... (?P<tld>[a-z]{2,4})
... '''
>>> S = pd.Series(data)
>>> S.str.extract(pattern, re.VERBOSE)
          user  domain  tld
Dave      dave  google  com
multiple   rob   gmail  com
none       NaN     NaN  NaN
>>> 
```

That's great, but sometimes we want to extract all matches in each element of the series. You can do that with `S.str.findall` but its result does not include the names specified in the capturing groups of the regular expression:

``` python
>>> S.str.findall(pattern, re.VERBOSE)
Dave                           [(dave, google, com)]
multiple    [(rob, gmail, com), (steve, gmail, com)]
none                                             NaN
dtype: object
>>> 
```

I propose the `S.str.extractall` method which returns a `Series` the same length as the subject `S`. Each element of the series is a `DataFrame` with a row for each match and a column for each group:

``` python
>>> result = S.str.extractall(pattern, re.VERBOSE)
>>> result[0]
   user  domain  tld
0  dave  google  com
>>> result[1]
    user domain  tld
0    rob  gmail  com
1  steve  gmail  com
>>> result[2]
Empty DataFrame
Columns: [user, domain, tld]
Index: []
>>> 
```

Before I write any more testing code, can we start a discussion about whether or not this is an acceptable design choice, in relation to the other functionality of pandas? @sinhrks @jorisvandenbossche @jreback @mortada since you seem to be discussing extract in #10103

Also do you have any ideas about how to get the result (a Series of DataFrames) to print more nicely? With my current fork we have

``` python
>>> result
Dave                   user  domain  tld
0  dave  google  com
multiple        user domain  tld
0    rob  gmail  com
1  s...
none        Empty DataFrame
Columns: [user, domain, tld]
I...
dtype: object
>>> 
```

In R the equivalent functionality is provided by the https://github.com/tdhock/namedCapture package (str_match_all_named returns a list of data.frames), and the resulting printout is readable because of the way that R prints lists:

``` r
> library(namedCapture)
> S <- c(
+   Dave='dave@google.com',
+   multiple='rob@gmail.com some text steve@gmail.com',
+   none=NA)
> pattern <- paste0(
+   "(?P<user>[a-z]+)",
+   "@",
+   "(?P<domain>[a-z]+)",
+   "[.]",
+   "(?P<tld>[a-z]{2,4})")
> str_match_all_named(S, pattern)
$Dave
     user   domain   tld  
[1,] "dave" "google" "com"

$multiple
     user    domain  tld  
[1,] "rob"   "gmail" "com"
[2,] "steve" "gmail" "com"

$none
<0 x 0 matrix>

> 
```

jreback commented:
never embed DataFrames within a Series!

instead you *could* simply return a multi-indexed frame

jreback commented:
```
In [13]: DataFrame({'user' : ['dave','rob','steve',np.nan],'domain' : ['google.com','gmail.com','gmail.com',np.nan]},index=pd.MultiIndex.from_tuples([('Dave',0),('multiple',0),('multiple',1),('none',0)])
   ....: )
Out[13]: 
                domain   user
Dave     0  google.com   dave
multiple 0   gmail.com    rob
         1   gmail.com  steve
none     0         NaN    NaN
```

jreback commented:
@tdhock further you would prob add a parameter on how to handle partial matches (e.g. the ``some`` and ``text`` fields), e.g. would you be greedy then nan-fill or skip unless all filled (e.g. what you are doing now)

tdhock commented:
Thanks for the suggestions @jreback. Now extractall returns a DataFrame with a MultiIndex:

```python
>>> import re
>>> import pandas as pd
>>> import numpy as np
>>> data_dict = {
...     'single': {
...         "Dave":'dave@google.com',
...         "Toby":'tdhock5@gmail.com',
...         "Maude":'maudelaperriere@gmail.com',
...         },
...     'multiple': {
...         "robAndSteve": 'rob@gmail.com some text steve@gmail.com',
...         "abcdef": 'a@b.com some text c@d.com and e@f.com',
...         },
...     'none': {
...         "missing":np.nan,
...         "empty":"",
...         },
...     }
>>> tuple_list = []
>>> subject_list = []
>>> for k1, d in data_dict.iteritems():
...     for k2, subject in d.iteritems():
...         k = (k1, k2)
...         tuple_list.append(k)
...         subject_list.append(subject)
... 
>>> index = pd.MultiIndex.from_tuples(tuple_list, names=("matches", "subject"))
>>> Si = pd.Series(subject_list, index)
>>> named_pattern = r'''
... (?P<user>[a-z0-9]+)
... @
... (?P<domain>[a-z]+)
... \.
... (?P<tld>[a-z]{2,4})
... '''
>>> iresult = Si.str.extractall(named_pattern, re.VERBOSE)
>>> iresult
                                 user  domain  tld
matches  subject                                  
single   Dave                    dave  google  com
         Maude        maudelaperriere   gmail  com
         Toby                 tdhock5   gmail  com
multiple robAndSteve              rob   gmail  com
         robAndSteve            steve   gmail  com
         abcdef                     a       b  com
         abcdef                     c       d  com
         abcdef                     e       f  com
>>> S = pd.Series(subject_list)
>>> result = S.str.extractall(named_pattern, re.VERBOSE)
>>> result
              user  domain  tld
0             dave  google  com
1  maudelaperriere   gmail  com
2          tdhock5   gmail  com
3              rob   gmail  com
3            steve   gmail  com
4                a       b  com
4                c       d  com
4                e       f  com
>>>
```
So then we can access all the matches for the subject string with rob and steve via:
```python
>>> iresult.loc["multiple", "robAndSteve"]
                       user domain  tld
matches  subject                       
multiple robAndSteve    rob  gmail  com
         robAndSteve  steve  gmail  com
>>> result.loc[3]
    user domain  tld
3    rob  gmail  com
3  steve  gmail  com
```
For subjects that have 0 matches I think it would be more consistent and user-friendly if the following would return a DataFrame with 0 rows rather than an exception. Is that possible using some index options?
```python
>>> result.loc[5]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/core/indexing.py", line 1198, in __getitem__
    return self._getitem_axis(key, axis=0)
  File "pandas/core/indexing.py", line 1342, in _getitem_axis
    self._has_valid_type(key, axis)
  File "pandas/core/indexing.py", line 1304, in _has_valid_type
    error()
  File "pandas/core/indexing.py", line 1291, in error
    (key, self.obj._get_axis_name(axis)))
KeyError: 'the label [5] is not in the [index]'
>>> iresult.loc["none", "empty"]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/core/indexing.py", line 1196, in __getitem__
    return self._getitem_tuple(key)
  File "pandas/core/indexing.py", line 709, in _getitem_tuple
    return self._getitem_lowerdim(tup)
  File "pandas/core/indexing.py", line 822, in _getitem_lowerdim
    result = self._handle_lowerdim_multi_index_axis0(tup)
  File "pandas/core/indexing.py", line 804, in _handle_lowerdim_multi_index_axis0
    raise e1
KeyError: 'none'
>>> 
```

tdhock commented:
I further propose the `extractiter` method which returns an iterator over DataFrames, one for each subject: one row for each match, one column for each group:

```python
>>> for df in Si.str.extractiter(named_pattern, re.VERBOSE):
...   print df
... 
   user  domain  tld
0  dave  google  com
              user domain  tld
0  maudelaperriere  gmail  com
      user domain  tld
0  tdhock5  gmail  com
    user domain  tld
0    rob  gmail  com
1  steve  gmail  com
  user domain  tld
0    a      b  com
1    c      d  com
2    e      f  com
Empty DataFrame
Columns: [user, domain, tld]
Index: []
Empty DataFrame
Columns: [user, domain, tld]
Index: []
>>> for df in Si.str.extractiter(named_pattern, re.VERBOSE):
...   print df["domain"]
... 
0    google
Name: domain, dtype: object
0    gmail
Name: domain, dtype: object
0    gmail
Name: domain, dtype: object
0    gmail
1    gmail
Name: domain, dtype: object
0    b
1    d
2    f
Name: domain, dtype: object
Series([], Name: domain, dtype: object)
Series([], Name: domain, dtype: object)
>>> 
```

jreback commented:
@tdhock let's just keep it straightforward for now
we don't support iterators like this so it would be a big API change

tdhock commented:
OK, in that case I deleted extractiter, and added some more tests and examples for extractall.

jreback commented on pandas/core/strings.py:
```diff
@@ -502,6 +502,85 @@ def f(x):
+    
+    >>> S.str.extractall("(?P<letter>[ab])(?P<digit>\d)")
+      letter digit
+    A      a     1
```
why would this NOT be a multi-index here? Having a duplicated index is not at all convenient.

tdhock replied:
not sure what you mean. Can you please clarify?

When the input Series is not multi-indexed, there is no reason the output DataFrame should be. This is the same as the behavior of the standard `extract` method:

```python
>>> from pandas import Series
>>> S = Series(["a1a2", "b1", "c1"], ["A", "B", "C"])
>>> S.str.extractall("(?P<letter>[ab])(?P<digit>\d)")
  letter digit
A      a     1
A      a     2
B      b     1
>>> S.str.extract("(?P<letter>[ab])(?P<digit>\d)")
  letter digit
A      a     1
B      b     1
C    NaN   NaN
>>> e_df = S.str.extract("(?P<letter>[ab])(?P<digit>\d)")
>>> e_df.index
Index([u'A', u'B', u'C'], dtype='object')
>>> e_df.keys()
Index([u'letter', u'digit'], dtype='object')
>>> ea_df = S.str.extractall("(?P<letter>[ab])(?P<digit>\d)")
>>> ea_df.index
Index([u'A', u'A', u'B'], dtype='object')
>>> ea_df.keys()
Index([u'letter', u'digit'], dtype='object')
>>> 
```

jreback replied:
``.extract`` returns a like-index object to the original. 

the proposed ``.extractall`` by definition will have duplicates of *some* of the index elements. This is very different. This by its ery nature *should* return a ``MultiIndex`` (or if the input has a multi-index), then add a level.

jreback commented on pandas/core/strings.py:
```diff
@@ -502,6 +502,85 @@ def f(x):
+def str_extractall(arr, pat, flags=0):
+    """
+    Find all non-overlapping groups in each 
+    subject string in the Series using regular expression pat.
```
explain how this is different than ``str.extract``

jreback commented:
can you rebase / squash and i'll take a look

tdhock commented:
OK @jreback this is my first time doing a rebase / squash. Did I do it correctly?

jreback commented:
yes that looks right

jreback commented:
- add to ``api.rst``
- add a note in ``v0.18.0.txt`` (use this PR number as the issue number), in Enhancement
- update ``text.rst``
- update test in ``test_categorical`` if needed (this implements ``.str`` as well)

jreback commented on pandas/core/strings.py:
```diff
@@ -502,6 +502,91 @@ def f(x):
+    
+    >>> S.str.extractall("(?P<letter>[ab])?(?P<digit>\d)")
+            letter digit
+      match             
```
I think this should have the ``Index`` name from the original Series (could be ``None``)

tdhock replied:
In fact the name of the index is taken from the original Series, and it could be None.

```
In [3]: S = Series(["a1a2", "b1", "c1"], ["A", "B", "C"])

In [5]: S.str.extractall("(?P<letter>[ab])?(?P<digit>\d)")
Out[5]: 
        letter digit
  match             
A 0          a     1
  1          a     2
B 0          b     1
C 0        NaN     1

In [6]: S.str.extractall("(?P<letter>[ab])?(?P<digit>\d)").index
Out[6]: 
MultiIndex(levels=[[u'A', u'B', u'C'], [0, 1]],
           labels=[[0, 0, 1, 2], [0, 1, 0, 0]],
           names=[None, u'match'])

In [7]: Sn = Series(["a1a2", "b1", "c1"], ["A", "B", "C"])

In [12]: Sn.str.extractall("(?P<letter>[ab])?(?P<digit>\d)")
Out[12]: 
              letter digit
capital match             
A       0          a     1
        1          a     2
B       0          b     1
C       0        NaN     1

```

Do you think I should add a name to the Series used in the docstring?

jreback replied:
not necessary, just make sure have a test for the name. This is now the default to check names (on series/frame comparisons)

tdhock replied:
OK, indeed there is already a test for a subject `Series` with a named index.

jreback commented on pandas/core/strings.py:
```diff
@@ -502,6 +502,91 @@ def f(x):
+    Indices with no matches will not appear in the result.
+
+    >>> S = Series(["a1a2", "b1", "c1"], ["A", "B", "C"])
+    >>> S.str.extractall("[ab](?P<digit>\d)")
```
show S here (and make all lower case ``s``)

tdhock replied:
OK

jreback commented on pandas/tests/test_strings.py:
```diff
@@ -642,6 +642,129 @@ def test_extract_single_series_name_is_preserved(self):
+            expected_tuples, expected_index, expected_columns)
+        computed_df = S.str.extractall(named_pattern, re.VERBOSE)
+        tm.assert_frame_equal(computed_df, expected_df)
+        # The index of the input Series should be used to construct
```
put blank lines between different tests here (or sections of tests), makes more readable

tdhock replied:
OK

jreback commented on pandas/core/strings.py:
```diff
@@ -502,6 +502,91 @@ def f(x):
+    regex = re.compile(pat, flags=flags)
+    # the regex must contain capture groups.
+    if regex.groups == 0:
+        raise ValueError("This pattern contains no groups to capture.")
```
need a tests for this error condition, also if only have 1 pattern, assert that this returns the same as ``.extract``

tdhock replied:
OK

jreback commented on pandas/core/strings.py:
```diff
@@ -502,6 +502,91 @@ def f(x):
+        first_col = columns[0]
+        name = first_col if first_col in regex.groupindex else None
+        result = Series(first_list, index, name=name)
+    else:
```
IIRC this is what the ``expand`` kw does (or maybe we should always return a ``DataFrame`` here?)

tdhock replied:
I see the `expand` keyword defined in https://github.com/pydata/pandas/pull/10103 as 

```
Or you can specify ``expand=False`` to return Series.
```

I agree that sometimes returning a `DataFrame` and sometimes returning a `Series` is confusing from a user perspective.

This design choice (return a `Series` if there is only one group) was made to be consistent with the current implementation of `extract`.

I'm not sure if the analogy with find/findall is appropriate:
- find takes a string (not a regex) and returns the index of the first match.
- findall takes a regex with capture groups, and returns the contents of those capture groups.

```python

In [3]: s = Series(["a1a2", "b1", "c1"], ["A", "B", "C"])

In [10]: s.str.find("a")
Out[10]: 
A    0
B   -1
C   -1
dtype: int64

In [11]: s.str.findall("a")
Out[11]: 
A    [a, a]
B        []
C        []
dtype: object


```

tdhock replied:
If it was my choice I would always return a `DataFrame`, in both `extract` and `extractall`!

jreback replied:
@tdhock I think I agree with you here.ok, let's always return a ``DataFrame``. This will entail changing ``.extract`` and putting in API breaking changes. @jorisvandenbossche ?

tdhock commented:
TODO update docstrings

tdhock commented:
TODOs
- [ ] double check api.rst
- [ ] add a note in v0.18.0.txt (use this PR number as the issue number), in Enhancement
- [ ] double check text.rst
- [ ] add extractall tests to test_categorical? (this implements .str as well) Is this needed? There are no tests for extract in test_categorical.

jreback commented:
.str functions are all tested in test--categorical - only the ones that need args are special cased

jreback commented on doc/source/text.rst:
```diff
@@ -205,6 +205,22 @@ and optional groups like
+   S = pd.Series(["a1a2", "b1", "c1"], ["A", "B", "C"])
+   S.str.extract("[ab](?P<digit>\d)")
+
+the ``extractall`` method (introduced in version 0.18)
```
use a ``versionadded`` tag here

tdhock replied:
OK

jreback commented on pandas/core/strings.py:
```diff
@@ -472,9 +472,17 @@ def str_extract(arr, pat, flags=0):
+    
+    if isinstance(arr, Index):
+        if 1 < regex.groups:
+            raise ValueError("only one regex group is supported with Index")
```
do you have a test for this?

tdhock replied:
Seems to me that there is no reason we should impose a limit of only one regex group for indices. This feature comes from https://github.com/pydata/pandas/issues/9980 which explains that they want extract to return an Index when it is called on an Index. In that case it seems like extract with one group should return an Index, and with several groups it should return a MultiIndex (not an error as in current implementation). what do you think?

tdhock replied:
In my opinion we should not return Index, even when it is called on an Index.

extract/extractall should always return DataFrame, and then if the user wants he can create an Index or MultiIndex.

Again this should be easier to maintain and easier for users to understand.

jreback commented on pandas/core/strings.py:
```diff
@@ -486,24 +494,97 @@ def f(x):
+    columns = [names.get(1 + i, i) for i in range(regex.groups)]
+    if len(arr) == 0:
+        result = DataFrame(columns=columns, dtype=object)
     else:
```
you don't need an else here (as the if returns)

tdhock replied:
OK

jreback commented on pandas/core/strings.py:
```diff
@@ -486,24 +494,97 @@ def f(x):
+    Series has exactly one match, extractall(pat).xs(0, level='match')
+    is the same as extract(pat).
+
+    Parameters
```
add a ``versionadded`` tag here

tdhock replied:
OK

jreback commented on pandas/tests/test_strings.py:
```diff
@@ -636,12 +636,200 @@ def check_index(index):
+        r = s.str.extract(r'(?P<letter>[a-z])')
+        e = DataFrame({"letter":['a', 'b', 'c']})
+        tm.assert_frame_equal(r, e)
+
```
on these test add a comment with this issue number (and an explanation if the title is not enough)

tdhock replied:
OK

jreback commented:
@tdhock need a big note in whatsnew API breaking changes to show off the prior / new behavior in ``.extract``

jreback commented on doc/source/text.rst:
```diff
@@ -205,6 +205,22 @@ and optional groups like
+
+Unlike ``extract`` (which returns only the first match),
+
+.. ipython:: python
```
There is also a Method Summary section, pls add ``.extracall`` there as well

tdhock replied:
OK

jreback commented on doc/source/text.rst:
```diff
@@ -205,6 +207,24 @@ and optional groups like
 
+.. _text.extractall:
+
+Unlike ``extract`` (which returns only the first match),
```
add a warning box about the pre-0.18.0 behavior or ``.extract``

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -20,6 +20,21 @@ Check the :ref:`API Changes <whatsnew_0180.api>` and :ref:`deprecations <whatsne
 ~~~~~~~~~~~~
 
+.. _whatsnew_0180.enhancements.extractall:
+
```
put this under a separate sub-section (e.g. 'string extraction changes' or something)

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -20,6 +20,21 @@ Check the :ref:`API Changes <whatsnew_0180.api>` and :ref:`deprecations <whatsne
+The ``.str.extractall`` method is added (:issue:`11386`).
+Unlike ``extract`` (which returns only the first match),
+
+.. ipython:: python
```
these need to be indented 2 spaces

tdhock replied:
not sure I understand. what exactly needs to be indented 2 spaces? In the other whatsnew/*.txt files,
- The python code blocks seem to be indented 3 spaces, and
- The text following `_whatsnew` blocks does not seem to be indented.

jreback replied:
so this is correct now, since we added a sub-section. If you didn't have a sub-section, then you would need to indent everything except the first line (e.g. including the ipython line)

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -20,6 +20,21 @@ Check the :ref:`API Changes <whatsnew_0180.api>` and :ref:`deprecations <whatsne
+.. ipython:: python
+
+   s = pd.Series(["a1a2", "b1", "c1"], ["A", "B", "C"])
+   s.str.extract("[ab](?P<digit>\d)")
```
show s here as well

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -86,13 +101,78 @@ In addition, ``.round()`` will be available thru the ``.dt`` accessor of ``Serie
+groups, finds the first match in each subject string, and returns the
+contents of the capture groups.
 
+**PRIOR** to 0.18.0, ``extract`` could return a ``Series``, ``Index``,
```
use the 'Previous Behavior'/'New Behavior' tags (see 0.17.0.txt for how this is formatted)

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -86,13 +101,78 @@ In addition, ``.round()`` will be available thru the ``.dt`` accessor of ``Serie
 
+**PRIOR** to 0.18.0, ``extract`` could return a ``Series``, ``Index``,
+or ``DataFrame``, depending on the subject and regular expression
+pattern. We concluded that this is confusing from a user perspective,
```
just say, 'This is confusing from a user perspective.....' (rather than We concluded)

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -86,13 +101,78 @@ In addition, ``.round()`` will be available thru the ``.dt`` accessor of ``Serie
+.. code-block:: python
+
+   In [3]: s = pd.Series(["a1", "b2", "c3"], ["A11", "B22", "C33"])
+   In [10]: one_group = "(?P<letter>[a-zA-Z])"
```
use a python block before this to define & show s; the code-block should just show that result

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -86,13 +101,78 @@ In addition, ``.round()`` will be available thru the ``.dt`` accessor of ``Serie
+
+.. ipython:: python
 
+   s = pd.Series(["a1", "b2", "c3"], ["A11", "B22", "C33"])
```
then don't need to re-define s again here (its the same)

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -86,13 +101,78 @@ In addition, ``.round()`` will be available thru the ``.dt`` accessor of ``Serie
 
+Previously, calling ``index.str.extract`` on a regex with exactly one
+capture group returned an ``Index``:
+
```
I think take all of these 3 example cases (e.g. Index/Series with 1 and more than 1 groups), and show examples in ``text.rst``

tdhock replied:
OK for the two examples about Index, but there is already an example for a Series with one group, which I did not copy to avoid repetition.

jreback replied:
ok that sound good. just making sure we have coverage of the cases that changed.

jreback commented on pandas/tests/test_strings.py:
```diff
@@ -558,15 +551,15 @@ def test_extract(self):
-        exp = Series([NA, NA, NA], dtype=object)
-        tm.assert_series_equal(result, exp)
+        exp = DataFrame([NA, NA, NA], dtype=object)
+        tm.assert_frame_equal(result, exp)
```
hmm, add a name to the Series which should be the name of the resultant column in the frame, yes?

tdhock replied:
Actually the resultant column name(s) should come from the capture group name(s). I added a test to make sure that extract gives the same output for an input series with and without a name.

jreback commented on pandas/tests/test_strings.py:
```diff
@@ -538,13 +538,6 @@ def test_extract(self):
-        # GH9980
-        # Index only works with one regex group since
-        # multi-group would expand to a frame
-        idx = Index(['A1', 'A2', 'A3', 'A4', 'B5'])
```
do you have a test with an Index with 1 capture group (and more than one)? (which should now work)

tdhock replied:
Yes.

tdhock replied:
"test_extract_index_one_two_groups"

jreback commented:
@tdhock looks pretty good, just some minor comments.

@jorisvandenbossche comments?

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -86,13 +102,101 @@ In addition, ``.round()`` will be available thru the ``.dt`` accessor of ``Serie
+
+.. code-block:: python
+
+   >>> s = pd.Series(["a1", "b2", "c3"], ["A11", "B22", "C33"])
```
this should be an executable block

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -86,13 +102,101 @@ In addition, ``.round()`` will be available thru the ``.dt`` accessor of ``Serie
+.. code-block:: python
+
+   >>> one_group = "(?P<letter>[a-zA-Z])"
+   >>> s.str.extract(one_group)
```
this is correct, but use ipython style blocks for the output

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -86,13 +102,101 @@ In addition, ``.round()`` will be available thru the ``.dt`` accessor of ``Serie
+
+New Behavior:
+
+.. code-block:: python
```
this can be an ipython block (use s from above)

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -86,13 +102,101 @@ In addition, ``.round()`` will be available thru the ``.dt`` accessor of ``Serie
 
+.. code-block:: python
 
+   >>> two_groups = "(?P<letter>[a-zA-Z])(?P<numbers>[0-9]+)"
```
use an ipython style output block (but still code-block)

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -86,13 +102,101 @@ In addition, ``.round()`` will be available thru the ``.dt`` accessor of ``Serie
 
+.. code-block:: python
 
+   >>> s.index.str.extract(two_groups)
```
make this an ipython block

jreback commented on pandas/core/strings.py:
```diff
@@ -415,36 +415,38 @@ def _get_single_group_name(rx):
     Returns
     -------
-    extracted groups : Series (one group) or DataFrame (multiple groups)
-        Note that dtype of the result is always object, even when no match is
```
maybe add this last sentence back (about the ``object`` dtype)

tdhock replied:
OK

jreback commented on pandas/core/strings.py:
```diff
@@ -486,24 +489,103 @@ def f(x):
-        result = np.array([f(val)[0] for val in arr], dtype=object)
-        name = _get_single_group_name(regex)
+    names = dict(zip(regex.groupindex.values(), regex.groupindex.keys()))
+    columns = [names.get(1 + i, i) for i in range(regex.groups)]
```
this is annoying, but prob should have a test that verifies when you have *some* named groups and others that are not named, you would get columns like ``Index(['foo',1,'bar',3])``.

jreback commented on pandas/core/strings.py:
```diff
@@ -486,24 +489,103 @@ def f(x):
+    columns = [names.get(1 + i, i) for i in range(regex.groups)]
+    if len(arr) == 0:
+        return DataFrame(columns=columns, dtype=object)
     else:
```
don't need the else as you are returning in the if (and its more clear IMHO)

jreback commented on pandas/core/strings.py:
```diff
@@ -486,24 +489,103 @@ def f(x):
+    A 0          a     1
+      1          a     2
+    B 0          b     1
+    C 0        NaN     1
```
show an example w/o named groups as well.

tdhock replied:
OK

tdhock commented:
I am adding a test for extractall when there are no matches at all. In that case it should return an Empty DataFrame with columns named according to group names.

tdhock commented on pandas/tests/test_strings.py:
```diff
@@ -558,50 +551,56 @@ def test_extract(self):
+    def test_extract_series(self):
+        # extract should give the same result whether or not the
+        # series has a name.
+        for series_name in None, "series_name":
```
here is the test that makes sure extract gives the same result for a series with and without a name.

tdhock commented on pandas/tests/test_strings.py:
```diff
@@ -635,13 +634,248 @@ def check_index(index):
+        # one un-named group.
+        r = s.str.extractall('(z)')
+        e = DataFrame(columns=[0])
+        tm.assert_frame_equal(r, e)
```
here are the tests that make sure extractall works even when there are no matches at all.

tdhock commented:
@jreback would you suggest any other changes?

jreback commented on doc/source/text.rst:
```diff
@@ -205,6 +217,52 @@ and optional groups like
 can also be used.
 
+When ``str.extract`` method is used on an ``Index``, it returns a
+``DataFrame`` with as many columns as capture groups.
```
and if the capture group has a ``?P<...>`` name then it will be named like this, otherwise, numbered for its position?

tdhock replied:
Yes.

tdhock replied:
Do I need to clarify the docs? I thought it was clear enough from the first examples without named groups, followed by the discussion of named groups.

jreback replied:
we put it on the doc-string, yes? I think it should be mentioned in text.rst (maybe in a ``::note``) as I imagine this could be a bit surprising / not obvious to a user.

tdhock replied:
it is in the examples of the doc-string but I will also add a specific description under Returns: "Any capture group names in regular expression pat will be used for column names; otherwise capture group numbers will be used." I also added that to text.rst.

jreback commented on doc/source/text.rst:
```diff
@@ -205,6 +217,52 @@ and optional groups like
+   s = pd.Series(["a1", "b2", "c3"], ["A11", "B22", "C33"])
+   s
+   s.index.str.extract("(?P<letter>[a-zA-Z])")
+   s.index.str.extract("(?P<letter>[a-zA-Z])(?P<numbers>[0-9]+)")
```
show w/o naming the 2nd capture group (as a 3rd example)

tdhock replied:
OK

jreback commented on doc/source/text.rst:
```diff
@@ -205,6 +217,52 @@ and optional groups like
+
+.. versionadded:: 0.18.0
+
+the ``extractall`` method (introduced in version 0.18) returns a
```
you don't need to say introduced in 0.18, that's what the ``versionadded`` tag does

tdhock replied:
OK

jreback commented on doc/source/text.rst:
```diff
@@ -205,6 +217,52 @@ and optional groups like
+
+the ``extractall`` method (introduced in version 0.18) returns a
+``DataFrame`` with a ``MultiIndex`` that contains all matches.
+
```
the matches

tdhock replied:
how about "every match" ? (to contrast extract, which returns only the first match)

jreback commented on doc/source/text.rst:
```diff
@@ -205,6 +217,52 @@ and optional groups like
+   s.str.extractall("[ab](?P<digit>\d)")
+
+The last level of the ``MultiIndex`` is named ``match`` and indicates
+the order in the subject. When each subject string in the Series has
```
use a second ``ipython`` group and put the 2nd sentence before that (e.g. ``When each subject string``)...easier to follow along like that

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -20,6 +20,22 @@ Check the :ref:`API Changes <whatsnew_0180.api>` and :ref:`deprecations <whatsne
 
+.. _whatsnew_0180.enhancements.extractall:
+
+The ``.str.extractall`` method is added (:issue:`11386`).
```
put a ``:ref:`` link to the documentation section

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -86,13 +102,84 @@ In addition, ``.round()`` will be available thru the ``.dt`` accessor of ``Serie
+
+The ``.str.extract`` method takes a regular expression with capture
+groups, finds the first match in each subject string, and returns the
+contents of the capture groups.
```
put a ``:ref:`` link to the docs here as well

tdhock replied:
OK

jreback commented:
@jorisvandenbossche can you have a look

jreback commented:
@sinhrks @TomAugspurger comments?

TomAugspurger commented:
Just gave a quick skim, looks like a great addition.

@tdhock @jreback, is breaking the API to always return a DataFrame a necessary part of this change? How much messier would your code have to get to support maybe returning an Index or Series? I agree that always returning a DataFrame is nice, but maybe not nice enough to warrant breaking API.

Could we deprecate returning an Index/Series for a release? In the code you could check the output shape a squeeze to a Series for now, and raise a FutureWarning in that case. Again I haven't looked closely, so you'll know better than me how much of a hassle this will be.

jreback commented:
@TomAugspurger the API change on ``str.extract`` is not necessary, but IMHO it is just simply confusing. Depdending on the data (e.g. whether you have 1 or more capture groups), you get an Index/Series or a DataFrame. Better to just be consistent and always return a DataFrame.

We could deprecate this, but that is just kicking the can and not actually providing an upgrade path.

TomAugspurger commented:
Agreed that the current behavior is confusing, and it bites me often. Always returning a DataFame is a better behavior.

> We could deprecate this, but that is just kicking the can and not actually providing an upgrade path.

The deprecation path would need to include a flag like `squeeze` that defaults to `True` for a release before flipping to default, and being removed further down the road. So we'll require people to change their code once before 0.19, but at least the have a chance to adapt over the 0.18 series.

I don't like introducing a keyword just for the deprecation cycle, so I'm really on the fence here. I just know that `extractall` is a widely used method. Would be curious to hear others' thoughts here.

jreback commented:
actually we use the ``expand`` flag for several other method, so not w/o precedent. I suppose we could introduce here with a deprecation warning (and default it)

TomAugspurger commented:
Yes, `expand=False` for now to match `str.split` would be better than `squeeze=True`. Is `expand` on `str.split` slated for an API change to flip to True? Would make sense to do those together, if we were wanting to flip that (I don't think I've ever intentionally kept `expand=False` on `str.split`).

jreback commented:
ok, so here's a proposal. How about adding ``expand=False`` in 0.18.0; then adding deprecation warnings for both ``.extract`` and ``.split`` (shown if its an ``Index/Series``) that going to change for 0.19.0.

@tdhock I think adding the parameter ``expand=False`` (and restoring the functionailiy for ``.extract``), then another PR for the deprecations.

TomAugspurger commented:
:+1:

jorisvandenbossche commented:
A question: as it is a bit difficult to track in the long discussion, can somebody summarize the reasoning to change the return type to a DataFrame? Or is the only reason that is 'is confusing'? (the reason given in the whatsnew docs)

As I personally don't find this so confusing. I rather find it clear: doing an extract on a Series (Series.str.extract) gives you a Series (1 group) or a DataFrame (more groups). Note that this does 
*not* depend on your data, but on the regular expression you provide to the function. 

Further, the fact that the extract on an Index no longer returns an Index but a DataFrame, I don't know if this is an improvement. To be consistent with the other methods, shouldn't it rather return a Multi-Index?

jreback commented:
@jorisvandenbossche I agree that an ``Index`` *should* return a ``MultiIndex`` (but this is not implemented, so for sure we should do that).

But, this is data dependent on the moment, which is very confusing. If you have a regex with 1 group you get a ``Series``, or 2 you get a ``DataFrame``. So if the passed regex is unknown to the writer at that time (say its a user variable), then you don't know what you are getting back.

We try to be very predictable, this IMHO is pretty unpredictable.

shoyer commented:
I agree that always returning a frame would be a better API. It's pretty common to use variables for regexs.

On Sat, Dec 5, 2015 at 7:47 AM, Jeff Reback <notifications@github.com>
wrote:

> @jorisvandenbossche I agree that an ``Index`` *should* return a ``MultiIndex`` (but this is not implemented, so for sure we should do that).
> But, this is data dependent on the moment, which is very confusing. If you have a regex with 1 group you get a ``Series``, or 2 you get a ``DataFrame``. So if the passed regex is unknown to the writer at that time (say its a user variable), then you don't know what you are getting back.
> We try to be very predictable, this IMHO is pretty unpredictable.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/pydata/pandas/pull/11386#issuecomment-162213238

jreback commented:
this overlaps with the ideas in #10103 , cc @sinhrks 

where we add in ``expand=False`` as the default (and discuss options for deprecations and such)

jreback commented:
@tdhock can you update when you have a chance

(e.g. add the ``expand`` parameter), so need to put back some of the original impl for ``.extract``. We will deprecate (but in another issue).

tdhock commented:
yes I can do that probably some time next week. (I was a bit busy with other projects this week.)

jreback commented:
@tdhock if you'd like to update as discussed (and rebase) would be great.

tdhock commented:
hey sorry for the delays Jeff. I've been taking time off work to spend time
with family. Happy Holidays by the way! I should have time to look at this
by the first week of Jan.

On Fri, Dec 25, 2015 at 7:52 PM, Jeff Reback <notifications@github.com>
wrote:

> @tdhock <https://github.com/tdhock> if you'd like to update as discussed
> (and rebase) would be great.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas/pull/11386#issuecomment-167268838>.
>

jreback commented:
@tdhock pls update when you can.

also run: ``git diff master | flake8 --diff`` and fix errors. We are going to shortly be enforcing PEP8 compliance.

tdhock commented:
Hey @jreback I added the `expand` parameter to `extract`. For the docstring should we describe the behavior of both `expand=True` and `expand=False`?

jreback commented:
yep

tdhock commented:
I also ran flake8 and fixed all the problems.

tdhock commented:
Any other changes you would suggest @jreback ?

jorisvandenbossche commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -235,6 +309,58 @@ Subtraction by ``Timedelta`` in a ``Series`` by a ``Timestamp`` works (:issue:`1
+   pd.NaT / 2
+   pd.NaT * np.nan
+
+``NaT`` defines more arithmetic operations with ``datetime64[ns]`` and ``timedelta64[ns]``.
```
Is this part meant to added here? (seems from another PR?)

tdhock replied:
I think these additions came in after I merged with current master (otherwise there were conflicts). Should I remove these additions or leave them?

jorisvandenbossche commented on pandas/core/strings.py:
```diff
@@ -505,6 +507,280 @@ def f(x):
+    no match is found.
+
+    Examples
+    --------
```
Are the examples needed here? Because it seems duplicate from the `str_extract` docstring, at that one is the only that users will see?

jorisvandenbossche replied:
But, great examples!!

tdhock replied:
I will remove them.

jorisvandenbossche commented on pandas/core/strings.py:
```diff
@@ -505,6 +507,280 @@ def f(x):
+        re module flags, e.g. re.IGNORECASE
+    expand : bool, default False
+        * If True, return DataFrame.
+        * If False, return Series/Index/DataFrame.
```
Can you add a `.. versionadded::` for the expand keyword

tdhock replied:
OK

jorisvandenbossche commented on pandas/core/strings.py:
```diff
@@ -505,6 +507,280 @@ def f(x):
+    A pattern with one group will return a DataFrame with one column.
+    Indices with no matches will not appear in the result.
+
+    >>> s = Series(["a1a2", "b1", "c1"], ["A", "B", "C"])
```
can you add `index=` here (seems clearer to me)

jorisvandenbossche commented on pandas/core/strings.py:
```diff
@@ -505,6 +507,280 @@ def f(x):
+    no match is found. If expand=True and pat has only one capture group,
+    then return a Series (if subject is a Series) or Index (if subject
+    is an Index).
+
```
Maybe also add a 'See also' in this docstring that refers to extractall (and vice versa there)

tdhock replied:
OK

jorisvandenbossche commented:
Some other questions (from looking at the related PR https://github.com/pydata/pandas/pull/10103 from @sinhrks )

- What do you think of using the default `expand=None` instead of `False`, as the current behaviour is not really a 'False' (in the sense as it is for eg `str.split`) as it expands or not depending on the groups. This leaves the possibility to implement a 'real' False later on (although that seems not that useful).
- Is it very difficult to include the index -> MultiIndex behaviour? (instead of creating a DataFrame). I think that is the behaviour we ideally want? (but just not implemented?) Because now we are introducing new behaviour that we want to change, which sounds a bit strange.

tdhock commented:
Thanks for the suggestions @jorisvandenbossche.
- `expand=None` seems reasonable, but are the True/None/False options something that pandas users would expect?
- I would argue that that behavior that we ideally want is expand(anything, expand=True) -> DataFrame (easiest to understand). Then if the user wants (s)he can create a MultiIndex explicitly.
- Currently index.expand(regex_with_more_than_one_group, expand=False) raises ValueError. This is the case I believe where we could implement returning a MultiIndex, but it seems like a theoretical use case (I wouldn't use it), so I would be hesitant to implement that.

jreback commented on doc/source/text.rst:
```diff
@@ -168,28 +168,34 @@ Extracting Substrings
+
+   In version 0.18.0, ``extract`` gained the ``expand`` argument. When
+   ``expand=False`` (default) it returns a ``Series``, ``Index``, or
+   ``DataFrame``, depending on the subject and regular expression
```
I though the plan was to make ``expand=None`` the default. Then you can provide a deprecation warning, and support ``expand=False`` (return Series,Index,or DataFrame as appropriate to the data), and have the default actually be ``expand=True`` going forward. This provides back-compat with a warning, and explicit behavior.

tdhock replied:
Before I write any code for `expand=None` could you explain the difference between it and `expand=False` ?

Also for `extractall` would it be OK to simply not implement the `expand` argument?

tdhock replied:
Is it just that `expand=None` gives a warning and `expand=False` does not?

jreback replied:
@tdhock that's exactly right. The purpose of setting it to ``None`` is to detect if its explictly set or not. The only valid values are ``True/False``, but we detect ``None``, show the warning, then set it to ``False``.

jreback replied:
we recently merged this: https://github.com/pydata/pandas/pull/11149, where we did exactly this. The original way of using it was incorrect, so we provide a warning to make users be explict about what behavior is needed, while the default does not warn.

jorisvandenbossche replied:
> that's exactly right. The purpose of setting it to None is to detect if its explictly set or not. The only valid values are True/False, but we detect None, show the warning, then set it to False.

Not necessarily I think. In the PR of @sinhrks (#10103), he implemented it as the following (copied from the docstring);

```
+    expand : None or bool, default None
+        * If None, return Series/Index (one group) or DataFrame/MultiIndex (multiple groups)
+        * If True, return DataFrame/MultiIndex expanding dimensionality.
+        * If False, return Series/Index.
```

Where both True and False are new behaviours, and `None` indicates the current behaviour

tdhock replied:
About the PR of @sinhrks (#10103),
```
+        * If False, return Series/Index.
```
I don't think that makes sense for a regex with multiple groups (unless we want to raise `ValueError` when `expand=False`).

I think it makes more sense to do as @jreback proposed (have `None` just be an alias for the default, with a warning).

jorisvandenbossche replied:
With `expand=False`, you get back a Series with lists if there are multiple groups. I agree that is not really useful in most cases IMO, but it is how `split` has been from the beginning (and for that reason `expand=True` was added)

jorisvandenbossche replied:
In any case, it makes the dimension of the output perfectly predictable: `expand=False` -> 1D return, `expand=True` -> 2D return, and not dependent on the content of the regex as is now the case.

jreback replied:
agree on that. That's the point I think.

tdhock commented:
By the way is there some way we could include a table in the docs to explain how the different outputs are related to the different inputs?

|  | 1 group | >1 group   |
|-----------------|---------|------------|
| Series          | Series  | DataFrame  |
| Index           | Index   | ValueError |

jreback commented:
looks at for example ``doc/source/io.rst``, search for ``csv-table``. and use that directive.

jorisvandenbossche commented:
> I would argue that that behavior that we ideally want is anything.extract(any_regex, expand=True) -> DataFrame (easiest to understand). Then if the user wants (s)he can create a MultiIndex explicitly.

I don't fully agree with this. For me it seems also easy to understand that for an index, you get an index as result? 
But, the two most important reasons I am in favor of returning a MultiIndex are:

- If you want a MultiIndex as result, you say you can explicitly create it. But, I think it is easier to the other way around: if you want a DataFrame, just convert your index to a series:

  ```
df.index.to_series().extract(..)
# or
df.reset_index()['index'].extract(..)
  ```
  instead of making a multi-index from your dataframe:

  ```
temp = df.index.extract(...)
pd.MultiIndex.from_arrays(temp.values, names=temp.columns)
  ```

* Consistency: this keyword was recently introduced first for [`split`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html#pandas.Series.str.split). And there the meaning is very clear: if expand=True, return DataFrame/MultiIndex, if expand=False return Series/Index. And I would very like to keep the meaning of this keyword consistent throughout the string methods.

> Currently index.extract(regex_with_more_than_one_group, expand=False) raises ValueError. This is the case I believe where we could implement returning a MultiIndex, but it seems like a theoretical use case (I wouldn't use it), so I would be hesitant to implement that.

As far as I can see, the return of a MultiIndex (but for `expand=True`) is already implemented in #10103, so maybe you can look there to copy some code? (or we can leave it for later, but then I would raise an error now, so we don't have to change the behaviour)

jorisvandenbossche commented:
And for the table, you can also use some formatting: see http://sphinx-doc.org/rest.html#tables (instead of the directive)

tdhock commented:
Thanks for the in-depth analysis @jorisvandenbossche. I agree with you, it does seem a bit harder to make a `MultiIndex` from a `DataFrame`.

However this PR is already way too big and so for now I would propose to leave `index.extract(regex_with_multiple_groups, expand=False)` -> `ValueError` and then maybe implement retuning a `MultiIndex` in a future PR.

jorisvandenbossche commented:
Note that I am also arguing for returning a Multi-Index for the case of `index.str.extract(expand=True)` (what now returns a DataFrame). So the MultiIndex return won't be implemented in this PR (which is totally understandable, this PR has already taken very long due to all discussions), I would also return a ValueError (or NotImplementedError) for that case (which would mean to always raise if it is called from an index)

jreback commented:
should we have @sinhrks limit #10103 and merge that first (limit in meaning dont' apply this to ``.get_dummies``)

tdhock committed:
ENH: extract(expand), extractall

tdhock commented:
I just fixed the merge conflicts with master, and proof-read the changes again. Would you suggest fixing anything else before merging?

jreback commented on doc/source/text.rst:
```diff
@@ -201,9 +207,92 @@ and optional groups like
+   pd.Series(['a1', 'b2', 'c3']).str.extract('[ab](\d)', expand=True)
+   pd.Series(['a1', 'b2', 'c3']).str.extract('[ab](\d)', expand=False)
+
+Calling ``index.str.extract`` on a regex with exactly one capture
```
I would say calling on an ``Index`` (e.g.don't use ``index.str.extract``) with ``.str.extract``. Then this language is the same as above.

tdhock replied:
OK

jreback commented on doc/source/text.rst:
```diff
@@ -201,9 +207,92 @@ and optional groups like
+expression will be used for column names; otherwise capture group
+numbers will be used.
+
+Extracting a regular expression with one group returns a DataFrame
```
where you use ``DataFrame`` always use double-backticks (or ``Series`` or ``Index``)

tdhock replied:
OK

jreback commented on doc/source/text.rst:
```diff
@@ -201,9 +207,92 @@ and optional groups like
+.. ipython:: python
+
+   s.index.str.extract("(?P<letter>[a-zA-Z])([0-9]+)", expand=True)
+   s.index.str.extract("(?P<letter>[a-zA-Z])([0-9]+)", expand=False)
```
either use a ``code-block`` here to, or I think you need a ``:okexception:`` under the ``ipython``

tdhock replied:
OK

jreback commented on doc/source/text.rst:
```diff
@@ -201,9 +207,92 @@ and optional groups like
+.. ipython:: python
+
+   pd.Series(['a1', 'b2', 'c3']).str.extract('[ab](\d)', expand=True)
+   pd.Series(['a1', 'b2', 'c3']).str.extract('[ab](\d)', expand=False)
```
since below you are going to show ``expand=True/False``, do these in the same order.

Further I think its ok to break these into separate ``ipython`` blocks (e.g. with your commentary in-between)

tdhock replied:
OK

jreback commented on doc/source/text.rst:
```diff
@@ -201,9 +207,92 @@ and optional groups like
++--------+---------+------------+
+|        | 1 group | >1 group   |
++--------+---------+------------+
+| Index  | Index   | ValueError |
```
we are defering ``expand=True`` on an ``Index`` -> ``MultiIndex`` for later, yes?

tdhock replied:
yes

jreback commented on pandas/core/strings.py:
```diff
@@ -505,6 +459,248 @@ def f(x):
+    str_extract(expand=True), and always returns a DataFrame.
+
+    """
+    from pandas.core.series import Series
```
``from pandas import Series, DataFrame, Index``

tdhock replied:
OK

jreback commented on pandas/core/strings.py:
```diff
@@ -413,58 +413,12 @@ def _get_single_group_name(rx):
 
 
-def str_extract(arr, pat, flags=0):
+def str_extract_original(arr, pat, flags=0):
```
maybe call this ``str_extract_data_dependent_expand`` (or if you have a better name, ``original`` doesn't do it for me)

jreback replied:
I would make this and ``str_extract_frame`` both private, e.g. ``_str_extract_frame``

tdhock replied:
OK how about `_str_extract_noexpand`?

jreback replied:
sure

jreback commented on pandas/core/strings.py:
```diff
@@ -505,6 +459,248 @@ def f(x):
+
+    """
+    from pandas.core.series import Series
+    from pandas.core.frame import DataFrame
```
same import patter here as above

tdhock replied:
OK

jreback commented on pandas/tests/test_strings.py:
```diff
@@ -630,24 +630,426 @@ def check_index(index):
-                      tm.makeDateIndex, tm.makePeriodIndex]:
+        i_funs = [
+            tm.makeStringIndex, tm.makeUnicodeIndex, tm.makeIntIndex,
+            tm.makeDateIndex, tm.makePeriodIndex
```
add ``tm.makeRangeIndex``

tdhock replied:
OK

jreback commented:
@tdhock really good!

I didn't see any tests asserting the ``FutureWarning`` for passing ``expand=None``. can you add?

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -136,6 +137,78 @@ New Behavior:
+   s.str.extract("(?P<letter>[a-zA-Z])", expand=False)
+   s.str.extract("(?P<letter>[a-zA-Z])", expand=True)
+
+Calling ``index.str.extract`` on a regex with exactly one capture
```
same comments as in ``text.rst``

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -278,6 +351,58 @@ Subtraction by ``Timedelta`` in a ``Series`` by a ``Timestamp`` works (:issue:`1
+``NaT`` defines more arithmetic operations with ``datetime64[ns]`` and ``timedelta64[ns]``.
+
+.. ipython:: python
+
```
you will need to rebase on master

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -136,6 +137,78 @@ New Behavior:
+``expand=False`` it returns a ``Series``, ``Index``, or ``DataFrame``,
+depending on the subject and regular expression pattern (same behavior
+as pre-0.18.0). When ``expand=True`` it always returns a
+``DataFrame``, which is more consistent and less confusing from the
```
give a little explanation of the requirement for passing ``expand`` (otherwise you will get the warning)

tdhock replied:
OK

tdhock commented:
I rebased with master and removed the duplications in the whatsnew file.

I ran the tests on my machine but I am getting many error which are unrelated to my PR

```
thocking@silene:~/pandas(extractall)$ nosetests pandas/tests/test_strings.py
.E..EE.EEE....E........E..EEE..EE..E......EEEEE....EEE...EEE.EE.E..E..EE......E.EE...
======================================================================
ERROR: test_capitalize (pandas.tests.test_strings.TestStringMethods)
----------------------------------------------------------------------

======================================================================
ERROR: test_title (pandas.tests.test_strings.TestStringMethods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/thocking/pandas/pandas/tests/test_strings.py", line 288, in test_title
    tm.assert_almost_equal(mixed, exp)
  File "testing.pyx", line 58, in pandas._testing.assert_almost_equal (pandas/src/testing.c:3342)
  File "testing.pyx", line 139, in pandas._testing.assert_almost_equal (pandas/src/testing.c:2287)
  File "/home/thocking/pandas/pandas/core/series.py", line 558, in __getitem__
    result = self.index.get_value(self, key)
  File "/home/thocking/pandas/pandas/indexes/base.py", line 1920, in get_value
    raise IndexError(key)
IndexError: 0

----------------------------------------------------------------------
Ran 85 tests in 0.369s

FAILED (errors=34)
```

jreback commented:
did you rebuild the extensions, e.g. ``make``?

tdhock commented:
thanks for the tip. After rebuilding the extensions all tests pass on my machine.

jreback commented on doc/source/text.rst:
```diff
@@ -201,9 +207,106 @@ and optional groups like
 .. ipython:: python
 
-   pd.Series(['a1', 'b2', '3']).str.extract('(?P<letter>[ab])?(?P<digit>\d)')
+   pd.Series(['a1', 'b2', '3']).str.extract('([ab])?(\d)')
```
make ``extract`` and ``extractall`` sub-sections (I think you might have to use ``^^^^``) as the sub-headings

tdhock replied:
OK

jreback commented on pandas/core/strings.py:
```diff
@@ -418,38 +418,129 @@ def _get_single_group_name(rx):
+
+    empty_row = [np.nan] * regex.groups
+
+    def f(x):
```
*might* worth it to separate out this ``f(x)`` to an external function factory to generate for ``_str_extract_noexpand`` and here to not repeat the function

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -136,6 +136,89 @@ New Behavior:
+perspective of a user (:issue:`11386`). Currently the default is
+``expand=None`` which gives a ``FutureWarning`` and uses
+``expand=False``. To avoid this warning, please explicitly specify
+``expand``.
```
show an example (maybe in a note), that shows the future warning you will get when you don't supply ``expand``

tdhock replied:
OK

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -136,6 +136,89 @@ New Behavior:
+
+The :ref:`.str.extract <text.extract>` method takes a regular
+expression with capture groups, finds the first match in each subject
+string, and returns the contents of the capture groups.
```
put this issue number (e.g. the PR number) at the end of this first sentence (as an issue number)

tdhock replied:
OK

jreback commented:
@tdhock just a couple of doc changes. ping when pushed and green and we'll merge.

tdhock commented on pandas/core/strings.py:
```diff
@@ -418,38 +418,123 @@ def _get_single_group_name(rx):
+    groups_or_na = _groups_or_na_fun(regex)
+
+    if regex.groups == 1:
+        result = np.array([groups_or_na(val)[0] for val in arr], dtype=object)
```
`groups_or_na(subject)` should be easier to understand than `f(subject)`

tdhock commented on pandas/core/strings.py:
```diff
@@ -418,38 +418,123 @@ def _get_single_group_name(rx):
+    from pandas.core.index import Index
+
+    regex = re.compile(pat, flags=flags)
+    groups_or_na = _groups_or_na_fun(regex)
```
To avoid repetition, `_groups_or_na_fun(regex)` is used in both `_str_extract_noexpand` and `_str_extract_frame`

tdhock commented:
OK @jreback  I think I have addressed all your concerns.

jreback commented:
@tdhock thanks! great PR! and you put up with our comments!

only last thing: http://pandas-docs.github.io/pandas-docs-travis/ will have the built docs (may take a bit of time as Travis is sometimes queued). This builds all docs & doc-strings. Have a look and pls issue a followup-PR if anything needs clarification / formatting.
